### PR TITLE
refactor: add a finalizer to close the client

### PR DESCRIPTION
**Pull request by yin1999**
Reference: #245

Different from the previous PR, this PR adds client checks to all exposed methods to make the behavior of all methods consistent.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -168,55 +190,50 @@ func (client *Client) SetLanguage(langs ...string) error {
 
 // DisableOutput ...
 func (client *Client) DisableOutput() error {
-	err := client.SetVariable(DEBUG_FILE, os.DevNull)
-
-	client.setVariablesToInitializedAPIIfNeeded()
```
Comment (yin1999) [inline client.go:173]: This method is called in `client.SetVariable()` method

File: client.go [hunk 2]
```diff
@@ -168,55 +190,50 @@ func (client *Client) SetLanguage(langs ...string) error {
...
-	return err
+	return client.SetVariable(DEBUG_FILE, os.DevNull)
 }
 
 // SetWhitelist sets whitelist chars.
 // See official documentation for whitelist here https://tesseract-ocr.github.io/tessdoc/ImproveQuality#dictionaries-word-lists-and-patterns
 func (client *Client) SetWhitelist(whitelist string) error {
-	err := client.SetVariable(TESSEDIT_CHAR_WHITELIST, whitelist)
-
-	client.setVariablesToInitializedAPIIfNeeded()
```
Comment (yin1999) [inline client.go:183]: see the above comment

File: client.go [hunk 3]
```diff
@@ -168,55 +190,50 @@ func (client *Client) SetLanguage(langs ...string) error {
...
-	return err
+	return client.SetVariable(TESSEDIT_CHAR_WHITELIST, whitelist)
 }
 
 // SetBlacklist sets blacklist chars.
 // See official documentation for blacklist here https://tesseract-ocr.github.io/tessdoc/ImproveQuality#dictionaries-word-lists-and-patterns
 func (client *Client) SetBlacklist(blacklist string) error {
-	err := client.SetVariable(TESSEDIT_CHAR_BLACKLIST, blacklist)
-
-	client.setVariablesToInitializedAPIIfNeeded()
```
Comment (yin1999) [inline client.go:193]: see the above comment

File: client.go [hunk 4]
```diff
@@ -82,12 +85,19 @@ func (client *Client) Close() (err error) {
 	// 		err = fmt.Errorf("%v", e)
 	// 	}
 	// }()
+	if client.api == nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (yin1999) [reply client.go]: Good point. The cases of `api == nil` would be (please correct this if there are any missing cases):

- The client is created by `client := &gosseract.Client{}` (in this case, the `SetFinalizer` is not called)
- The client has been closed before (we have removed the finalizer in the previous `Close()` call)
- The `gosseract.NewClient()` has failed to initialize api (the `C.Create()` call in `NewClient()` function return a nil pointer)

For the last case, we can choose to remove the finalizer, or leave it as it is (the "no-op" `Close()` method will be called before it is GCed, while there is no any allocated cgo memory).

I believe the frequency of the last case occurs should be very low.

**Comment by otiai10** [comment]
@yin1999 can you take a look at the failing test on mingw?
If you have Docker on your machine, it's easy to debug

File: client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (otiai10) [reply client.go]: From readability perspective, I think it's good idea to call `SetFinalizer( ,nil)` anyway, anycase in `Close` method.
Let me know your thought

[the remark this answers is not in the document]
Comment (yin1999) [reply client.go]: It shouldn't have a huge impact on performance, I'll handle it along with the failing test :)

**Review by otiai10 — APPROVED** [review]

**Pull request merged**